nifi-api/src/main/java/org/apache/nifi/annotation/documentation/DeprecationNotice.java

```diff
@@ -0,0 +1,51 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.nifi.annotation.documentation;
+
+import org.apache.nifi.components.ConfigurableComponent;
+
+import java.lang.annotation.Documented;
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Inherited;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+
+/**
+ * Annotation that can be applied to a {@link org.apache.nifi.processor.Processor Processor},
+ * {@link org.apache.nifi.controller.ControllerService ControllerService}, or
+ * {@link org.apache.nifi.reporting.ReportingTask ReportingTask} in order to
+ * warn about the deprecation of the component. The deprecation warning is informational only
+ * and doesn't affect the processor run time behavior in any way
+ */
+@Documented
+@Target({ElementType.TYPE})
+@Retention(RetentionPolicy.RUNTIME)
+@Inherited
+public @interface DeprecationNotice {
+    Class<? extends ConfigurableComponent>[] value() default {};
+
+    String[] classNames() default {};
+
+    String reason() default "";
+
+
+    // The following are still not utilised but implemented
+    // for future use.
+    String deprecationTargetVersion() default "";
```

Comment:
@mcgilman - I placed it in there to address a commend from the list (I think it was Andy's suggestion to have it) but given I think it will take a while before we can truly agree on deprecating something by version X, it may be reasonable to remove it for now.

From the other hand, unless you foresee this going into the UI we may well adjust the documentation generation logic to take this field in consideration. 

Happy with whatever you prefer.